design/tls-client-verification.md

```diff
@@ -0,0 +1,125 @@
+# External client certificate validation (mutual TLS authentication)
+
+Status: Draft
+
+This document outlines how to add support for authentication of external
+clients by validating their client certificates.
+
+## Goals
+
+- Allow Contour to be used to protect the backend services from access of
+  unauthorized external clients.
+- Allow authentication of external clients by having Envoy validate that
+  client certificates are signed by trusted CA (Certificate Authority)
+- Allow configuration of trusted CA certificate(s) for validating the client
+  certificates.
+
+## Non Goals
+
+- Fine grained authorization on level of individual request
+  (example of non-goal: only client X can access resource Y on backend
+  service Z)
+
+## Background
+
+In TLS (and HTTPS) often only the client authenticates the server.  TLS
+supports optional authentication of client which is referred to as
+*mutual TLS authentication* or *mTLS*.  Mutual TLS authentication is typically
+used in machine-to-machine (M2M) communication e.g. to protect sensitive REST
+APIs.  The application acting as TLS client authenticates itself towards the
+TLS server by using x509 certificate and by providing proof of posession of
+the corresponding private key.
+
+
+## High-Level Design
+
+Envoy supports following options for certificate based client authentication:
+
+1. Verify that a chain of trust can be established from the presented client
+   certificate to the configured trusted root CA certificate (validation of
+   certificate)
+2. In addition to 1, verify that the subject alternative name of the client is
+   one of the names listed in the configuration (validation of client identity)
+3. Verify that the client certificate hash or subject public key hash matches
+   with configured hash (hash pinning)
+
+Option 1) is proposed to be implemented in this document.  It is sufficient
+for the very simplest mutual TLS authentication use cases.
+
+At a high level following CRD change is proposed:
+
+*  A new record `clientValidation` is added in `spec.virtualhost.tls`
+   in the `HTTPProxy`.
+
+*  The `clientValidation` contains parameter `caSecret` which is a reference
```

Comment:
What is the justification for a new key, do you think we'll have to add additional keys under `clientValidation` in the future?

Reply:
I wish that in future we could add verification of subject alternative name of the client (option 2. in the list above).

Reply:
We should use `UpstreamValidation` (with a new name) to describe both upstream and downstream TLS validation policy.